src/LibraryManager.Vsix/Shared/VsHelpers.cs

```diff
@@ -336,5 +372,138 @@ public static Project GetDTEProjectFromConfig(string file)
+                filesToAdd.RemoveRange(0, countToDelete);
+            }
+
+            return true;
+        }
+
+        private static async Task<bool> AddProjectItemsInBatchAsync(IVsHierarchy vsHierarchy, List<string> filePaths, Action<string, LogLevel> logAction, CancellationToken cancellationToken)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            IVsProjectBuildSystem bldSystem = vsHierarchy as IVsProjectBuildSystem;
+
+            try
+            {
+                if (bldSystem != null)
+                {
+                    bldSystem.StartBatchEdit();
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var vsProject = (IVsProject)vsHierarchy;
+                VSADDRESULT[] result = new VSADDRESULT[filePaths.Count()];
+
+                vsProject.AddItem(VSConstants.VSITEMID_ROOT,
+                            VSADDITEMOPERATION.VSADDITEMOP_LINKTOFILE,
+                            string.Empty,
+                            (uint)filePaths.Count(),
+                            filePaths.ToArray(),
+                            IntPtr.Zero,
+                            result);
+
+                foreach (string filePath in filePaths)
+                {
+                    logAction.Invoke(string.Format(Resources.Text.LibraryAddedToProject, filePath.Replace('\\', '/')), LogLevel.Operation);
+                }
+            }
+            catch (Exception)
+            {
+                logAction.Invoke(string.Format(Resources.Text.FailToAddLibrariesToProject), LogLevel.Operation);
+
+                return false;
+            }
+            finally
+            {
+                if (bldSystem != null)
+                {
+                    bldSystem.EndBatchEdit();
+                }
+            }
+
+            return true;
+        }
+
+        private static async Task<bool> DeleteProjectItemsInBatchAsync(IVsHierarchy hierarchy, IEnumerable<string> filePaths, Action<string, LogLevel> logAction, CancellationToken cancellationToken)
+        {
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            IVsProjectBuildSystem bldSystem = hierarchy as IVsProjectBuildSystem;
+            HashSet<ProjectItem> folders = new HashSet<ProjectItem>();
+
+            try
+            {
+                if (bldSystem != null)
+                {
+                    bldSystem.StartBatchEdit();
+                }
+
+                foreach (string filePath in filePaths)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    ProjectItem item = DTE.Solution.FindProjectItem(filePath);
+
+                    if (item != null)
+                    {
+                        ProjectItem parentFolder = item.Collection.Parent as ProjectItem;
+                        folders.Add(parentFolder);
```

Comment:
It only deletes one level up. We need to add logic to determine how far up in the tree we can delete these empty folders